Commitment Discount Appendix (#528)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/appendix/commitment_discounts.md

# Commitment Discounts

> A commitment discount is a billing discount that offers reduced rates on preselected SKUs in exchange for an obligated usage or spend amount over a predefined term. Commitment discounts typically consist of a set of purchase and usage records within cost and usage datasets.

Reviewer: Would love this definition to not include the word `discount` in the first line. Something like:

```suggestion
A commitment discount is a rate reduction on preselected SKUs in exchange for an obligated usage or spend amount over a predefined term. Commitment discounts typically consist of a set of purchase and usage records within cost and usage datasets.
```

Author: I'm going to keep this as is to maintain consistency with this pr, the new `CommitmentDiscunt*` columns, and the FinOps terminology page, so actually need to add the missing word, "model".

<img width="657" alt="image" src="https://github.com/user-attachments/assets/63c8fc92-7551-43ee-b730-4e12c0128fd3">

## Purchasing

While customers are bound to the term of a commitment discount, cloud-service providers, or CSPs, offer various payment options before and/or during the term:

* *Upfront* - The commitment discount is paid in full before the term begins

> * *Monthly* - The commitment discount is paid monthly over the term.

Reviewer: Is `Monthly` the right term here?  There may be other grains we want to include here.  Suggest using `Recurring` or similar instead.

Author: Recurring works.

> * *Hybrid* (1/2 Upfront & 1/2 Monthly) - Half of the commitment discount is paid in advance and half is paid monthly over the term.

Reviewer: This needs to be genericized as `partial` or something similar, so as to include partial upfront that isn't exactly half.

> For example, if a customer buys a spend-based commitment discount for 1 year, with a $1.00 hourly commitment, and pays with the hybrid option, the commitment discount's payment consists of both:

Reviewer: This is the first mention of the term `spend-based`.  Given that this page is the explanation of a commitment discount, it should be written with the assumption that the reader brings no previous understanding with them.  This term deserves a definition earlier if you are going to use it here.

Author: I think by that logic, we would also need add definitions for:
- billing period
- charge period
- amortization
- term
...

Other than redefining `commitment discount` in the first sentence, I would rather link the glossary to the appendix rather than regurgitate definitions in 2 places. 

Thoughts?

# Commitment Discounts

> A commitment discount is a billing discount model that offers reduced rates on preselected SKUs in exchange for an obligated usage or spend amount over a predefined term. Commitment discounts typically consist of a set of purchase and usage records within cost and usage datasets.

Reviewer 2: ```suggestion
A commitment discount is a billing discount model that offers reduced rates on preselected SKUs in exchange for an obligated usage or spend amount over a predefined term. Commitment discounts typically consist of purchase and usage records within cost and usage datasets.
```

> Usage-based commitment discounts obligate a customer to a total quantity of time of a preselected term, typically measured in "Hours".  In some cases, usage-based commitment discounts allow similar resources of different size classifications to relatively benefit from commitment discount rates.  This is typically known as "size flexibility", and is determined by multiplying the number of actual hours consumed by a resource by its predetermined normalization factor to derive the number of "Normalized Hours" consumed.

Reviewer: While I am not willing to hold up the approval of this PR, I do think that this document will strongly benefit from either 1) presenting an example of how size flexibility surfaces in the data, or 2) explicitly stating that a deeper explanation of size flexibility is beyond the scope of this document.

Author: Agree, would like to iterate to that goal too.

Author: Size Flexibility Appendix: https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/535

> Spend-based commitment discounts obligate a customer to a predetermined amount of spend over a preselected term, typically denoted by the invoice's selected billing currency (ex: "USD").

Reviewer 3: When we discussed the other two commitment utilization-related columns, we decided 

- to be as vague as possible about commitment flexibility 
- and to avoid mentioning the billing currency (in the case of spend-based commitments). 
 
For the same reason, I suggest omitting these two chapters and incorporating the mention of commitment flexibility into the first chapter.

Author: Chris to rephrase general directions to point towards specific examples

## Purchasing

> While customers are bound to the term of a commitment discount, cloud-service providers offer some or all of the following payment options before and/or during the term:

Reviewer 2: ```suggestion
While customers are bound to the term of a commitment discount, providers offer some or all of the following payment options before or during the term:
```

* *Upfront* - The commitment discount is paid in full before the term begins.
* *Recurring* - The commitment discount is paid on a repeated basis, typically over each billing period of the term.

> * *Partial* - Some of the commitment discount is paid before the term begins, and the rest is paid repeatedly over the term.

Reviewer 3: Personally, I prefer using the terms All-Upfront, No-Upfront, and Partial-Upfront for payment models to avoid potential confusion with the Recurring Charge Frequency.
As for the descriptions, I suggest we focus on mentioning One-Time and/or Recurring purchases.

## Commitment Discounts in FOCUS

Within the FOCUS specification, the following examples demonstrate how a commitment discount would appear across various payment and usage scenarios.

### Purchase Rows

All commitment discount purchases appear with a positive `BilledCost`, `PricingCategory` as "Committed", and the commitment discount's id populating both the `ResourceId` and `CommitmentDiscountId` value. Upfront purchases appear as a single record also with `ChargeCategory` as "Purchase", `ChargeFrequency` as "One-Time", and the total quantity and units for commitment discount's term as `CommitmentDiscountPurchasedQuantity` and `CommitmentDiscountUnit`, respectively.

> Recurring purchases are allocated across all corresponding charge periods of the term when `ChargeCategory` is "Purchase", `ChargeFrequency` is "One-Time", and `CommitmentDiscountPurchasedQuantity` and `CommitmentDiscountUnit` are reflected only for that charge period.

Reviewer 3: Suggestion based on Scenario 2: Recurring

```suggestion
Recurring purchases are allocated across all corresponding charge periods of the term when `ChargeCategory` is "Purchase", `ChargeFrequency` is "Recurring", and `CommitmentDiscountPurchasedQuantity` and `CommitmentDiscountUnit` are reflected only for that charge period.
```

Reviewer 3: Regarding the editorial guidelines, which format are we supposed to use for column IDs — for instance, `ChargeFrequency` or ChargeFrequency? 
Based on https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/501, I assume it's the latter.

Author: Will incorporate styling guidelines across all pr's after merging content.

> Recurring purchases are allocated across all corresponding charge periods of the term when `ChargeCategory` is "Purchase", `ChargeFrequency` is "One-Time". and the commitment discount's purchase quantity, `CommitmentDiscountPurchasedQuantity`, and units, `CommitmentDiscountUnit` are reflected only for that charge period.

Reviewer: Would prefer to add consistency in one of the following ways:

```suggestion
Recurring purchases are allocated across all corresponding charge periods of the term when `ChargeCategory` is "Purchase", `ChargeFrequency` is "One-Time", and `CommitmentDiscountPurchasedQuantity` and `CommitmentDiscountUnit` are reflected only for that charge period.
```

```suggestion
Recurring purchases are allocated across all corresponding charge periods of the term when the charge category (`ChargeCategory`) is "Purchase", charge frequency (`ChargeFrequency`) is "One-Time", and the commitment discount's purchase quantity (`CommitmentDiscountPurchasedQuantity`) and units (`CommitmentDiscountUnit`) are reflected only for that charge period.
```

        "ChargePeriodEndDate": "2024-01-01 00:00:00Z",
        "ChargeCategory": "Purchase",
        "ChargeFrequency": "One-Time",
        "PricingCategory": "Committed",
        "ResourceId": "<commitment-discount-id>",
        "BilledCost": 4380.00,
        "EffectiveCost": 0.00,
        "CommitmentDiscountId": "<commitment-discount-id>",
        "CommitmentDiscountPurchasedQuantity": 4380.00,
        "CommitmentDiscountUnit": "USD"
    },
    {
        "BillingPeriodStartDate": "2023-01-01 00:00:00Z",
        "BillingPeriodEndDate": "2023-02-01 00:00:00Z",
        "ChargePeriodStartDate": "2023-01-01 00:00:00Z",
        "ChargePeriodEndDate": "2023-01-01 01:00:00Z",
        "ChargeCategory": "Purchase",
        "ChargeFrequency": "Recurring",
        "PricingCategory": "Committed",
        "ResourceId": "<commitment-discount-id>",
        "BilledCost": 0.50,
        "EffectiveCost": 0.00,
        "CommitmentDiscountId": "<commitment-discount-id>",
        "CommitmentDiscountPurchasedQuantity": 0.50,
        "CommitmentDiscountUnit": "USD"
    },

    /* ... 743 more recurring purchase records for the billing period ... */
]
```

### Usage Rows

> Amortization of commitment discounts occur the same way regardless of how one or more purchases are made.  The same usage-based or spend-based amount is applied evenly across all charge periods and potentially allocated to eligible resources.  Continuing with the same commitment discount example, a one-year, spend-based commitment discount with a $1.00 hourly commitment, 4 types of scenarios can occur during a charge period:

Reviewer: This section should remind the reader that there's only one resource in play in the example.
